### PR TITLE
[ADD] portal: manage api keys in portal

### PR DESCRIPTION
**Pull request by beledouxdenis**
Add the possibility for portal users to manage API keys in the frontend portal interface.

### Implementation choices
 - Display the API keys in the portal in debug mode only (as in the back-end for regular users, in their profile)
 - Allow for portal users to manage their API keys if the system parameter `portal.allow_api_keys` is set. ~~This is therefore a hidden feature except if you know this parameter. I do not see a way to enable the feature (to set this system parameter) from the settings (`res.config.settings`) without adding a column in database, which is forbidden in a stable, and this targets stable version 15.0. Which is not ideal at all. We do not like hidden features.
 An alternative would be to enable this feature even without this system parameter. So portal users would be able to manage API keys as soon as they go to debug mode, which seem acceptable to me, but this is a change of behavior in stable **by default**. An opinion on this is welcome.~~
 A settting in `res.config.settings` is added in order to enable or disable the feature (to add/remove the system parameter) through a compute field, to avoid adding a new database column in standard. See attached screenshot for illustration of its look.
The setting appear only in debug mode.
- I mimicked the flow, the wording and the look and feel from the back-end. And to mimic it perfectly, there is a touchy point: in the back-end, when you hit the "New API key" button, the password confirmation is asked directly, meaning before you need to enter the description of the key. This is thanks to the fact the button calls the method `api_key_wizard` on which the decorator `check_identity` is set.  
https://github.com/odoo/odoo/blob/5a87db453fcfd750934d1a1d62c00d93ad01c0e1/odoo/addons/base/models/res_users.py#L1650-L1651
Here in the front-end, there is not really a need to call this method to be able to create an API key. But, in order to mimic the flow as in the backend, this method is called with an `rpc` just so it asks the password confirmation before displaying the dialog to enter the key description. The result of the rpc call to `api_key_wizard` is left unused.
The flow is illustrated with the below screenshots. The screenshots are in the right order to understand the steps of the flow.
- Concerning the table displaying the keys in the portal, naturally I would have used the table `t-call="portal.portal_table"`, which display a nice rounded white table. This is the table used for other tables for documents in the portal (invoices, sale orders, etc.). See for example https://github.com/odoo/odoo/blob/5a87db453fcfd750934d1a1d62c00d93ad01c0e1/addons/account/views/account_portal_templates.xml#L35
But in this case, the table just under to display the trusted devices doesn't use it, and has a different look and feel. In order to be uniform and have the same look and feel for the table, I used the same table and classes than in this table for the trusted devices. See the screenshot below. Maybe using `t-call="portal.portal_table` on both these tables could be part of another commit in master.
 - In this API keys table, I choose to show the scope, as in the back-end view. Though, I do not really see the point, as the scopes are currently never used, and therefore this column will always be empty. This is therefore a bit of a shame to display a column we know will always be empty. Maybe we could remove it, and remove it as well from the back-end view in another dedicated commit.
- The sentences from the back-end views are copy/pasted in templates for the front-end views. It's a bit disappointing. Though this was already the case for the password confirmation view.  
See for instance
https://github.com/odoo/odoo/blob/5a87db453fcfd750934d1a1d62c00d93ad01c0e1/odoo/addons/base/views/res_users_views.xml#L44
https://github.com/odoo/odoo/blob/5a87db453fcfd750934d1a1d62c00d93ad01c0e1/addons/portal/static/src/xml/portal_security.xml#L4

To be uniform, and as I am working in a stable version, I chose this way, despite the copy/paste we like to avoid when possible. Though, one another way could be to move the code in the module `auth_totp_portal` parsing and fixing the back-end views to be used for the front-end, to avoid this copy-paste.
https://github.com/odoo/odoo/blob/5a87db453fcfd750934d1a1d62c00d93ad01c0e1/addons/auth_totp_portal/static/src/js/totp_frontend.js#L225-L231
But I do not think this something to do in stable 15.0.

### Screenshots
image
image
image
image
image

### CODE REVIEW

File: odoo/addons/base/models/res_users.py [hunk 1]
```diff
@@ -1753,7 +1753,8 @@ class APIKeyDescription(models.TransientModel):
     @check_identity
     def make_key(self):
         # only create keys for users who can delete their keys
-        if not self.user_has_groups('base.group_user'):
+        if not self.user_has_groups('base.group_user') and \
+           not tools.str2bool(self.env['ir.config_parameter'].sudo().get_param('portal.allow_api_keys')):
```
Comment (xmo-odoo) [inline odoo/addons/base/models/res_users.py:1757]: Maybe just `get_param`? That seems to be what many of the other settings do, and `set_param(key, False)` removes it if it exists.

File: addons/portal/controllers/portal.py [hunk 2]
```diff
@@ -204,6 +204,9 @@ def account(self, redirect=None, **post):
     def security(self, **post):
         values = self._prepare_portal_layout_values()
         values['get_error'] = get_error
+        values['allow_api_keys'] = tools.str2bool(
+            request.env['ir.config_parameter'].sudo().get_param('portal.allow_api_keys')
+        )
```
Comment (xmo-odoo) [inline addons/portal/controllers/portal.py:209]: Probably just `bool` or nothing at all? (see note below)

**Review by xmo-odoo — COMMENTED** [review]
> Allow for portal users to manage their API keys if the system parameter portal.allow_api_keys is set. This is therefore a hidden feature except if you know this parameter. I do not see a way to enable the feature (to set this system parameter) from the settings (res.config.settings) without adding a column in database, which is forbidden in a stable, and this targets stable version 15.0. Which is not ideal at all. We do not like hidden features.

Computed fields don't work at all in settings? It would require implementing the toggle by hand which is not great but...

>  Here in the front-end, there is not really a need to call this method to be able to create an API key. But, in order to mimic the flow as in the backend, this method is called with an rpc just so it asks the password confirmation before displaying the dialog to enter the key description. The result of the rpc call to api_key_wizard is left unused.

Maybe add a comment to the JS to explain that bit?

> To be uniform, and as I am working in a stable version, I chose this way, despite the copy/paste we like to avoid when possible. Though, one another way could be to move the code in the module auth_totp_portal parsing and fixing the back-end views to be used for the front-end, to avoid this copy-paste.

There it was done because the TOTP flow has its own objects and is quite complicated and it was not clear that everything was correct, so duplicating the flow was seen as risky (if there was a critical security issue in what I'd done and we forgot to fix one side).

It was a pain in the ass, would not recommend. And that was all JS from a button, here it's more complicated because stuff gets copied from a backend view (a form view) to a frontend template (a qweb view), I'm not even sure we can easily share code between those.

**Comment by beledouxdenis** [comment]
> Computed fields don't work at all in settings? It would require implementing the toggle by hand which is not great but...

Will have a look but last time I had a look it wasn't so obvious.

> Maybe add a comment to the JS to explain that bit?

Done

File: addons/portal/static/src/js/portal.js [hunk 3]
```diff
@@ -155,6 +156,75 @@ publicWidget.registry.portalSearchPanel = publicWidget.Widget.extend({
...
+                            }
+                        }
+                    ],
+                }).open();
+            }}, {text: _t('Discard'), close: true}],
+        }).open();
+    }
+});
+
+publicWidget.registry.RevokeTrustedDeviceButton = publicWidget.Widget.extend({
```
[the remark this answers is not in the document]
Comment (beledouxdenis) [reply addons/portal/static/src/js/portal.js]: Indeed, it's the result of a copy/paste :p. I renamed it :D

**Comment by beledouxdenis** [comment]
> Computed fields don't work at all in settings? It would require implementing the toggle by hand which is not great but...
It's done : )
image

**Comment by xmo-odoo** [comment]
> > Computed fields don't work at all in settings? It would require implementing the toggle by hand which is not great but...
> 
> It's done : ) image

nice

**Comment by beledouxdenis** [comment]
robodoo r+

**Pull request closed without merging**